src/core/net/ip6.cpp

```diff
@@ -934,29 +936,58 @@ otError Ip6::HandleExtensionHeaders(Message &    aMessage,
     return error;
 }
 
-otError Ip6::HandlePayload(Message &aMessage, MessageInfo &aMessageInfo, uint8_t aIpProto)
+void Ip6::HandlePayload(Message &          aMessage,
+                        MessageInfo &      aMessageInfo,
+                        uint8_t            aIpProto,
+                        Message::Ownership aMessageOwnership)
 {
-    otError error = OT_ERROR_NONE;
+    otError  error = OT_ERROR_NONE;
+    Message *message;
+
+    switch (aMessageOwnership)
+    {
+    case Message::kTakeCustody:
+        message = &aMessage;
+        break;
+
+    case Message::kCopyToUse:
+        VerifyOrExit((message = aMessage.Clone()) != nullptr, error = OT_ERROR_NO_BUFS);
+        break;
+
+    default:
```

Comment:
We can consider removing the `default` case here  (mainly  since the `Ownership` is an `enum` type with two enumerator values and we get a warning/error if a enumerator case value is not covered in `switch()`.

Reply:
Yes, but the compiler doesn't buy there are only two cases. It complains message may not be initialized. I am fine with initialize it.